### ISSUE TITLE
Is the image input during the training process the original image and the binary image?

### ISSUE THREAD

**Issue by lisongyu111** [khdlr/HED-UNet#3]

**Comment by khdlr** [comment]
Yes, you got that right.

The `img` and `target` parameters to the `full_forward` method used during training are exactly the original image (`img`) and the binary target segmentation (`target`).

Cheers, Konrad

**Comment by lisongyu111** [comment]
Thanks for answering! Where can I put the original picture in the folder in the program? Where shall I put the binary picture in the file?

**Comment by lisongyu111** [comment]
Thanks for answering! Where are the original pictures and binary pictures in the code?

**Comment by khdlr** [comment]
You can download the INRIA data here: https://project.inria.fr/aerialimagelabeling/

Here's the code that reads in the dataset: https://github.com/khdlr/HED-UNet/blob/master/deep_learning/utils/data.py

**Comment by lisongyu111** [comment]
Thanks Reply! What pictures are put in the AerialImageDataset and scenes texts in the code?

**Comment by khdlr** [comment]
The `AerialImageDataset` is the only folder that is being used. `scenes` is not used anymore, I will update the code and delete the `get_batch` function. Thanks for catching my error!

**Comment by lisongyu111** [comment]
Can you show me how you created the data file? I want to train my own data set.

**Comment by khdlr** [comment]
The dataset was not created by me, I just wrote the code that loads the data at training time.

What you need to do to train on your own data, is to implement a [`torch.utils.data.Dataset`](https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset) that returns pairs of image and ground truth annotations.

Then you can change the `get_dataset` function in `data_loading.py` to use your custom dataset instead of the pre-configured InriaDataset.

Let me know if that helps!

**Comment by lisongyu111** [comment]
I have 256*256 original images and binary images of aerial cities here, but I don’t know how to train this net

**Comment by khdlr** [comment]
Okay, so you'll subclass `torch.utils.data.Datset` like this

```python
class MyCustomDataset(torch.utils.data.Dataset):
  def __init__(self):
    # Do whatever initialization you need:
    self.images  = <list_of_image_paths>
    self.targets = <list_of_binary_annotations>

  def __getitem__(self, index):
    # "load_image" can be imageio.imread for example
    image  = load_image(self.images[index])
    target = load_image(self.targets[index])
    return image, target
  
  def __len__(self):
    return len(self.images)
```

Then you can simply change `data_loading.py` to use `MyCustomDataset` instead of `InriaDataset`.

Hope that helps.

**Comment by lisongyu111** [comment]
Thank you

**Comment by lisongyu111** [comment]
Hello, I need your help, I am having difficulty training the model.The training results are all 0.I think my data should be put in the wrong way. My data is a 256*256 picture, what size picture did you use in the lab 64*64.Hope you can answer me.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "khdlr/HED-UNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年9月14日(星期二) 下午5:17
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [khdlr/HED-UNet] Is the image input during the training process the original image and the binary image? (#3)





 
Okay, so you'll subclass torch.utils.data.Datset like this
 class MyCustomDataset(torch.utils.data.Dataset):   def __init__(self):     # Do whatever initialization you need:     self.images  = <list_of_image_paths&gt;     self.targets = <list_of_binary_annotations&gt;   def __getitem__(self, index):     # "load_image" can be imageio.imread for example     image  = load_image(self.images[index])     target = load_image(self.targets[index])     return image, target      def __len__(self):     return len(self.images)
 
Then you can simply change data_loading.py to use MyCustomDataset instead of InriaDataset.
 
Hoper that helps.
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by khdlr** [comment]
That is odd. I used fairly large pictures for training (768x768), but 256x256 should work.

Can you confirm that your masks are valid? What is the output if you add `print(torch.unique(target))` to the `full_forward` function in `train.py`?

**Comment by lisongyu111** [comment]
Thank you for your reply. What files are the original and two-difference diagrams of your data set respectively? I don't think my data set has been built correctly.



发自我的iPhone


------------------ Original ------------------
From: Konrad Heidler ***@***.***&gt;
Date: Sat,Sep 25,2021 4:52 PM
To: khdlr/HED-UNet ***@***.***&gt;
Cc: lisongyu111 ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [khdlr/HED-UNet] Is the image input during the training process the original image and the binary image? (#3)





 
That is odd. I used fairly large pictures for training (768x768), but 256x256 should work.
 
Can you confirm that your masks are valid? What is the output if you add print(torch.unique(target)) to the full_forward function in train.py?
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by khdlr** [comment]
Not entirely sure what you mean by original / two-difference diagrams.

Another issue might be data scaling - what is the output when you add `print('img', torch.min(img), torch.mean(img), torch.max(img))` in `full_forward`?

**Comment by lisongyu111** [comment]
Thanks, I encountered a small problem.







------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "khdlr/HED-UNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年9月25日(星期六) 下午5:31
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [khdlr/HED-UNet] Is the image input during the training process the original image and the binary image? (#3)





 
Not entirely sure what you mean by original / two-difference diagrams.
 
Another issue might be data scaling - what is the output when you add print('img', torch.min(img), torch.mean(img), torch.max(img)) in full_forward?
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by BinitPandit94** [comment]
As input, you have given the images and their segmented mask. I want to know what ground truth you have considered while modeling the edge detection part. It seems Sobel Kernel has been used for edge calculation (but in the paper, it has been written that your results are compared with those). Please highlight the Edge Detection part. Thank You.

**Comment by khdlr** [comment]
At training time, we compute the edge GT by applying Sobel to the segmentation GT for cases where no edge GT is available. As these segmentation masks are valued in {0,1}, we can recover perfect edge masks.

The "Sobel" comparison in the paper is referring to applying Sobel directly to the imagery at test time, which will generate much worse edge masks.
